docs/puzzles/day21.md

```diff
@@ -1,11 +1,312 @@
 import Solver from "../../../../website/src/components/Solver.js"
 
-# Day 21
+# Day 21: Dirac Dice
+by @sjrd
 
 ## Puzzle description
 
 https://adventofcode.com/2021/day/21
 
+## Modeling and parsing the input
+
+There are no more surprises here, so let us directly see the code:
+
+```scala
+type Cell = Int // from 0 to 9, to simplify computations
+
+case class Player(cell: Cell, score: Long)
+
+type Players = (Player, Player)
+
+def parseInput(input: String): Players =
+  val lines = input.split("\n")
+  (parsePlayer(lines(0)), parsePlayer(lines(1)))
+
+def parsePlayer(line: String): Player =
+  line match
+    case s"Player $num starting position: $cell" =>
+      Player(cell.toInt - 1, 0L)
+```
+
+The only thing worth nothing is that we use numbers 0 to 9 for the cells (the "spaces") instead of 1 to 10.
+The only reason is that it simplifies computations for wrapping around.
+
+## The deterministic die
+
+For the first play with the deterministic die, we have to return the score of the losing player, while keeping tabs on how many times the die was rolled.
+We model the deterministic die as an instance of a class `DeterministicDie`:
+
+```scala
+final class DeterministicDie {
+  var throwCount: Int = 0
+  private var lastValue: Int = 100
+
+  def nextResult(): Int =
+    throwCount += 1
+    lastValue = (lastValue % 100) + 1
+    lastValue
+}
+```
+
+An instance of that class keeps track of what its last roll was, and of how many times it was thrown.
+
+## Playing with the deterministic die
+
+We use the deterministic die in a tail-recursive function `playWithDeterministicDie` to play out a full game:
+
+```scala
+@tailrec
+def playWithDeterministicDie(players: Players, die: DeterministicDie): Long =
+  val diesValue = die.nextResult() + die.nextResult() + die.nextResult()
+  val player = players(0)
+  val newCell = (player.cell + diesValue) % 10
+  val newScore = player.score + (newCell + 1)
+  if newScore >= 1000 then
+    players(1).score
+  else
+    val newPlayer = Player(newCell, newScore)
+    playWithDeterministicDie((players(1), newPlayer), die)
+```
+
+In that function, it is always `players(0)`'s turn.
+In the tail-recursive call, we swap out the two players, so that the next player to play always comes first.
+
+Other than that, the function is a direct translation from the rules of the game in the puzzle description:
+
+1. Throw the die 3 times, and add up the values
+2. Advance the player, wrapping around every 10 cells
+3. Compute the new score
+4. Detect the winning condition and return the loser's score, or continue with the next player
+
+If you are not familiar with modular arithmetics (anymore?), it may not be obvious that the computation `(player.cell + diesValue) % 10` is correct.
```

Comment:
The `(anymore ?)` kind of assumes that most people learned and forgot about modular arithmetics. But we should not assume any background in mathematics from our readers.

```suggestion
If you are not familiar with modular arithmetics, it may not be obvious that the computation `(player.cell + diesValue) % 10` is correct.
```

Reply:
That was the point of the question mark, to say that it *may* be "anymore" for some people, and not at all for other people. But I could also remove it altogether.